core/src/main/java/com/linecorp/armeria/common/multipart/MultipartDecoder.java

```diff
@@ -98,16 +120,213 @@ public long demand() {
     @Override
     public void subscribe(Subscriber<? super BodyPart> subscriber, EventExecutor executor,
                           SubscriptionOption... options) {
-        decoded.subscribe(subscriber, executor, options);
+        requireNonNull(subscriber, "subscriber");
+        requireNonNull(executor, "executor");
+        requireNonNull(options, "options");
+
+        final MultipartSubscriber multipartSubscriber = new MultipartSubscriber(subscriber, executor);
+        if (!delegatedSubscriberUpdater.compareAndSet(this, null, multipartSubscriber)) {
+            // Avoid calling method on late multipartSubscriber.
+            // Because it's not static, so it will affect MultipartDecoder.
+            SubscriberUtil.failLateSubscriber(executor, subscriber, delegatedSubscriber.subscriber);
+            return;
+        }
+        decoded.subscribe(multipartSubscriber, executor, options);
     }
 
     @Override
     public void abort() {
-        decoded.abort();
+        abort(AbortedStreamException.get());
     }
 
     @Override
     public void abort(Throwable cause) {
+        requireNonNull(cause, "cause");
+        if (delegatedSubscriber == null) {
+            final MultipartSubscriber abortedSubscriber = new MultipartSubscriber(
+                    AbortingSubscriber.get(cause), EventLoopGroups.directEventLoop());
+            delegatedSubscriberUpdater.compareAndSet(this, null, abortedSubscriber);
+        }
         decoded.abort(cause);
     }
+
+    BodyPartPublisher onBodyPartBegin() {
+        return new BodyPartPublisher();
+    }
+
+    void requestUpstreamForBodyPartData() {
+        @Nullable
+        final MultipartSubscriber delegatedSubscriber = this.delegatedSubscriber;
+        if (delegatedSubscriber != null) {
+            delegatedSubscriber.requestUpstreamForBodyPartData();
+        }
+    }
+
+    final class BodyPartPublisher extends DefaultStreamMessage<HttpData> {
+        @Override
+        protected void onRequest(long n) {
+            // Because whenConsumed will run in the same thread(called by onRequest) after looping the existing
+            // queue.(onRequest & event notification in whenConsumed will run in the same executor specified
+            // at subscribe)
+            // So if there is no change, it means there is no buffered data processed.
```

Comment:
I'm not sure I understood the intention of this comment block. Could you elaborate? It'd be even nice if you can rephrase it a little bit.

Reply:
Let me recheck my comment here